python/lsst/daf/butler/core/logging.py

```diff
@@ -416,10 +412,17 @@ def from_raw(cls, serialized: Union[str, bytes]) -> ButlerLogRecords:
             return cls.parse_raw(serialized)
 
         # Filter out blank lines -- mypy is confused by the newline
-        # argument to split().
-        newline = "\n" if isinstance(serialized, str) else b"\n"
-        records = [ButlerLogRecord.parse_raw(line) for line in isplit(serialized, newline)  # type: ignore
-                   if line]
+        # argument to isplit() [which can't have two different types
+        # simultaneously] so we have to duplicate some logic.
```

Comment:
It was complaining that Union[str, bytes] for the separator can't work for a `str` parameter that is being split because the `T` type can't then work out what the return type is. With the rewrite it knows that serialized is the same type as newline so can know what isplit returns. Fundamentally if a variable at any point has a Union type it can't be converted to a non-Union type without being assigned to a new variable.

Reply:
> Fundamentally if a variable at any point has a Union type it can't be converted to a non-Union type without being assigned to a new variable.

I've definitely seen mypy "collapse" a union down to the remaining type(s) when one of them is rejected in other contexts  (e.g. if you `assert x is not None`, it can tell that `x` is `T` in `Optional[T`]).  But I guess there's not much value in trying to figure out why it couldn't do that here.

Reply:
Agreed; fine as-is.